Source/Diffable Collection/FetchedCollectionDiffableDataSource.swift

```diff
@@ -0,0 +1,134 @@
+	public var isUpdatingAutomatically: Bool = true {
+		didSet {
+			guard isUpdatingAutomatically else { return }
+			applyPendingSnapshot()
+		}
+	}
+
+	private let controller: NSFetchedResultsController<NSFetchRequestResult>
+	private let dataSource: FetchedDiffableDataSource
+	private lazy var internalSnapshot: NSDiffableDataSourceSnapshot<FetchedDiffableSection, FetchedDiffableItem> = .init() // non-modifiable snapshot to maintain data integrity, this snapshot stays in sync with the database
+	private var pendingSnapshot: NSDiffableDataSourceSnapshot<FetchedDiffableSection, FetchedDiffableItem>? // store pending changes while `isUpdatingAutomatically` is set to `false`
+	private weak var delegate: FetchedCollectionDiffableDataSourceDelegate?
+
+	// MARK: - Lifecycle
+
+	/// Creates a diffable data source based on a `NSFetchedResultsController` instance.
+	public init(controller: NSFetchedResultsController<NSFetchRequestResult>, dataSource: FetchedDiffableDataSource, delegate: FetchedCollectionDiffableDataSourceDelegate) {
+		self.controller = controller
+		self.dataSource = dataSource
+		self.delegate = delegate
+		super.init()
+		commonInit()
+	}
+
+	private func commonInit() {
+		setDelegate()
+		performFetch()
+	}
+
+	// MARK: - Controller management
+
+	/// Sets the delegate of the NSFetchedResultsController instance.
+	private func setDelegate() {
+		controller.delegate = self
+	}
+
+	/// Executes the fetch request of the NSFetchedResultsController instance.
+	private func performFetch() {
+		do {
+			try controller.performFetch()
+		} catch let error {
+			assertionFailure("Error performing controller fetch: \(error)")
+		}
+	}
+
+	// MARK: - NSFetchedResultsControllerDelegate
+	public func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChangeContentWith snapshot: NSDiffableDataSourceSnapshotReference) {
+		var itemsBeforeChange = internalSnapshot.itemIdentifiers
+
+		internalSnapshot.deleteSections(internalSnapshot.sectionIdentifiers)
+		let snapshotWithPermanentIDs = obtainPermanentIDs(for: snapshot, in: controller.managedObjectContext)
+		let typedSnapshot = snapshotWithPermanentIDs as NSDiffableDataSourceSnapshot<String, NSManagedObject>
+		typedSnapshot.sectionIdentifiers.forEach { sectionIdentifier in
+			let section: FetchedDiffableSection = .init(identifier: sectionIdentifier)
+			let items: [FetchedDiffableItem] = typedSnapshot.itemIdentifiers(inSection: sectionIdentifier).map { .init(item: $0) }
+			if (isHidingEmptySections && !items.isEmpty) || !isHidingEmptySections {
+				internalSnapshot.appendSections([section])
+				internalSnapshot.appendItems(items, toSection: section)
+			}
+		}
+
+		var externalSnapshot = internalSnapshot // modifiable snapshot to be displayed
+		delegate?.contentChangeWillBeApplied(snapshot: &externalSnapshot)
+
+		var itemsAfterChange = externalSnapshot.itemIdentifiers
+		var itemsToReconfigure = Array(Set(itemsBeforeChange).intersection(Set(itemsAfterChange)))
```

Comment:
This allows a user to modify incoming changes before they are displayed, right? From sections, to items' ordering, to the items themselves.

Doesn't that seem extremely dangerous though? We're talking about modifying MOs on the main thread, without saving, although a save might be triggered somewhere else, making these changes permanent.